src/core/thread/mle.cpp

```diff
@@ -3359,12 +3367,23 @@ bool Mle::IsMeshLocalAddress(const Ip6::Address &aAddress) const
 
 Router *Mle::GetParent(void)
 {
-    if ((!mParent.IsStateValidOrRestoring()) && (mParentCandidate.GetState() == Neighbor::kStateValid))
+    return &mParent;
```

Comment:
While trying to attach a new parent candidate as sleepy child, the devcie would send data poll. `MeshForwader` will use  the `GetParent()` to get destination for the data poll messages. This causes us to data poll to go to the incorrect/original parent. 

How about this?

In `GetParent()` we check if the `ParentCandiate` is in valid state, and if so we return `ParentCandiate`, otherwise we return `mParent`  (basically giving preference to a valid `mParentCandidate` over `mParent`).

We set the `ParentCandiate` state to valid from `SendChildIdRequest()` which is the time we would want to start data polling for "Child ID response" from new parent candidate. And then update `mParent = mParentCandidate` once we get the response and reset the parent candidate.

Reply:
This PR already introduces `Mle::GetParentCandidate()` that is used in `MeshForwarder::GetDirectTransmission()` for the `Message::kTypeMacDataPoll` case.

Reply:
Hmm... Indeed... You are correct... An internal test failed though (we saw child sending data poll to original/old parent)...